database/migrations/2021_06_01_181900_refactor_timestamps_anew.php

```diff
@@ -0,0 +1,563 @@
+	// "America/Argentina/Buenos_Aires" (both have 30 letters)
+	private const TZ_NAME_MAX_LENGTH = 31;
+	// All constants related to the Photos relation
+	private const PHOTOS_TABLE_NAME = 'photos';
+	private const PHOTO_TAKEN_AT_COL_NAME = 'taken_at';
+	private const PHOTO_TAKEN_AT_TZ_COL_NAME = 'taken_at_orig_tz';
+	private const PHOTO_TAKESTAMP_COL_NAME = 'takestamp';
+	// All constants related to the Configs relation
+	private const CONFIGS_TABLE_NAME = 'configs';
+	private const CONFIG_KEY_COL_NAME = 'key';
+	private const CONFIG_VALUE_COL_NAME = 'value';
+	private const CONFIG_RANGE_COL_NAME = 'type_range';
+	// All constants related to the configuration of Photo sorting (PS)
+	private const CONFIG_PS_KEY = 'sorting_Photos_col';
+	private const CONFIG_PS_VALUE_OLD2NEW = ['takestamp' => 'taken_at'];
+	private const CONFIG_PS_VALUE_NEW2OLD = ['taken_at' => 'takestamp'];
+	private const CONFIG_PS_RANGE_OLD = 'id|takestamp|title|description|public|star|type';
+	private const CONFIG_PS_RANGE_NEW = 'id|taken_at|title|description|public|star|type';
+	// All constants related to the configuration of Album sorting (AS)
+	private const CONFIG_AS_KEY = 'sorting_Albums_col';
+	private const CONFIG_AS_VALUE_OLD2NEW = [
+		'min_takestamp' => 'min_taken_at',
+		'max_takestamp' => 'max_taken_at',
+	];
+	private const CONFIG_AS_VALUE_NEW2OLD = [
+		'min_taken_at' => 'min_takestamp',
+		'max_taken_at' => 'max_takestamp',
+	];
+	private const CONFIG_AS_RANGE_OLD = 'id|title|description|public|max_takestamp|min_takestamp|created_at';
+	private const CONFIG_AS_RANGE_NEW = 'id|title|description|public|max_taken_at|min_taken_at|created_at';
+
+	/**
+	 * Run the migration.
+	 */
+	public function up()
+	{
+		try {
+			$this->fixPagesTable();
+			$this->upgradeORMSystemTimes();
+			$this->upgradePhotos();
+			$this->upgradeConfiguration();
+		} catch (\Exception $e) {
+			echo $e->getTraceAsString();
+			throw $e;
+		}
+	}
+
+	/**
+	 * Reverse the migration.
+	 */
+	public function down()
+	{
+		try {
+			$this->downgradeORMSystemTimes();
+			$this->downgradePhotos();
+			$this->downgradeConfiguration();
+		} catch (\Exception $e) {
+			echo $e->getTraceAsString();
+			throw $e;
+		}
+	}
+
+	/**
+	 * Fixes a buggy row in the table Pages.
+	 *
+	 * This bug was introduced by the migration `CreatePagesTable` on
+	 * 2019-02-21 11:43:56.
+	 * Normally, the Eloquent ORM framework ensures that the columns
+	 * `created_at` and `updated_at` are not null and are set correctly
+	 * if one calls {@link \Illuminate\Database\Eloquent\Model::save()}.
+	 * However, the migration manually inserted a row, by-passed the
+	 * Eloquent ORM layer and did not set these columns.
+	 */
+	protected function fixPagesTable(): void
+	{
+		DB::table('pages')
+			->where('title', '=', 'gallery')
+			->where('link', '=', '/gallery')
+			->whereNull('created_at')
+			->update([
+				'created_at' => '2019-02-21 11:43:56',
+				'updated_at' => '2019-02-21 11:43:56',
```

Comment:
I don't mind. I thought that it would be more correct to use the date/time which would have been used by the original migration which created that entry, but missed to set the dates back then. But actually I don't care.

What speaks for using the current date/time?

Reply:
Yeah, ideally we would set it to the time when the original migration ran. But we don't know when it was. For "old" users it may have been 2019, but for users who are just starting, it will have been "five seconds ago".

Reply:
After I had to learn that there are actually people who use this feature and create more entries manually (--> @ildyria), I incorporated @kamil4 suggestion, because we cannot know when those entries were created.